core/game.js

```diff
@@ -135,42 +131,48 @@ export class Game {
     this.mouse_delta.y -= e.movementY;
   }
 
+  frame(frame_time) {
+    if (this.running) {
+      window.requestAnimationFrame(frame_time => this.frame(frame_time));
+    }
+
+    if (frame_time > this.last_tick + this.tick_delta) {
+      const accumulated_delta = frame_time - this.last_tick;
+      const ticks_qty = Math.floor(accumulated_delta / this.tick_delta);
+      this.perform_ticks(ticks_qty);
+      this.render();
+    }
+  }
+
   perform_ticks(ticks_qty) {
     // Mouse delta is measured since the last time this.tick was run.  If there
     // is more than one tick to perform we need to scale the delta down to
     // maintain consistent movement.
     this.mouse_delta.x /= ticks_qty;
     this.mouse_delta.y /= ticks_qty;
 
-    for(var i=0; i < ticks_qty; i++) {
-      this.last_tick = this.last_tick + this.tick_length;
-      this.update(this.last_tick);
+    for (let i = 0; i < ticks_qty; i++) {
+      this.last_tick = this.last_tick + this.tick_delta;
+      this.update();
     }
 
     // Reset the mouse delta.
     this.mouse_delta.x = 0;
     this.mouse_delta.y = 0;
   }
 
-  update(tick_time) {
-    Object.keys(this.actions).forEach(action => {
-      if (this.actions[action]) {
-        this.actions[action](tick_time);
-      }
-    });
-
-    this.entities.forEach((entity) => entity.update(this.tick_length));
-
-    this.camera.update(this.tick_length);
+  update() {
+    this.emit('tick', this.last_tick);
```

Comment:
I wonder if we should perhaps rename `update` to `tick`, or the `tick` event to `update`?  That's a huge _I wonder_; I'm happy to leave this as-is.